wgpu-core/src/device/mod.rs

```diff
@@ -5331,10 +5353,54 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         &self,
         buffer_id: id::BufferId,
         range: Range<BufferAddress>,
-        op: resource::BufferMapOperation,
-    ) -> Result<(), resource::BufferAccessError> {
+        op: BufferMapOperation,
+    ) -> Result<(), BufferAccessError> {
+        // User callbacks must not be called while holding buffer_map_async_inner's locks, so we
+        // defer the error callback if it needs to be called immediately (typically when running
+        // into errors).
+        if let Err((op, err)) = self.buffer_map_async_inner::<A>(buffer_id, range, op) {
+            let status = match &err {
+                &BufferAccessError::Device(_) => BufferMapAsyncStatus::ContextLost,
+                &BufferAccessError::Invalid | &BufferAccessError::Destroyed => {
+                    BufferMapAsyncStatus::Invalid
+                }
+                &BufferAccessError::AlreadyMapped => BufferMapAsyncStatus::AlreadyMapped,
+                &BufferAccessError::MapAlreadyPending => BufferMapAsyncStatus::MapAlreadyPending,
+                &BufferAccessError::MissingBufferUsage(_) => {
+                    BufferMapAsyncStatus::InvalidUsageFlags
+                }
+                &BufferAccessError::UnalignedRange
+                | &BufferAccessError::UnalignedRangeSize { .. }
+                | &BufferAccessError::UnalignedOffset { .. } => {
+                    BufferMapAsyncStatus::InvalidAlignment
+                }
+                &BufferAccessError::OutOfBoundsUnderrun { .. }
+                | &BufferAccessError::OutOfBoundsOverrun { .. } => {
+                    BufferMapAsyncStatus::InvalidRange
+                }
+                _ => BufferMapAsyncStatus::Error,
+            };
+
+            op.callback.call(status);
+
+            return Err(err);
+        }
+
+        Ok(())
+    }
+
+    fn buffer_map_async_inner<A: HalApi>(
+        &self,
+        buffer_id: id::BufferId,
+        range: Range<BufferAddress>,
+        op: BufferMapOperation,
+    ) -> Result<(), (BufferMapOperation, BufferAccessError)> {
         profiling::scope!("Buffer::map_async");
 
+        // Note: We need the callback to be called in all error cases, however it must be called
+        // outside of buffer_map_async_inner, so that we don't hold any locks. Ensure that any
+        // early path that doesn't lead to a pending map state sets closure_to_resolve accordingly.
```

Comment:
This comment is out of date. I don't think we need it at all anymore, since the types essentially require us to hand back `op` on failure. Perhaps the best place for the comment is on the function itself, explaining why it has an interesting return type.